src/Butler/Auth/OIDC.hs

```diff
@@ -0,0 +1,219 @@
+    pure OIDCEnv{..}
+  where
+    OIDCClientID clientId = opClientID
+    OIDCClientSecret clientSecret = opClientSecret
+
+data OIDCState = OIDCState
+    { randomT :: Text
+    , uri :: Maybe Text
+    }
+    deriving (Generic, Show)
+
+instance FromJSON OIDCState
+
+instance ToJSON OIDCState
+
+-- | Generate a random fixed size string of 42 char base64 encoded
+genRandomB64 :: IO ByteString
+genRandomB64 = B64.encode . hash <$> genRandom
+
+-- | Generate a random fixed size string of 1024 Bytes
+genRandom :: IO ByteString
+genRandom = do
+    g <- Random.newStdGen
+    let (bs, _ng) = genByteString 1024 g
+    pure bs
+
+mkSessionStore :: OIDCEnv -> Maybe O.State -> Maybe Text -> O.SessionStore IO
+mkSessionStore OIDCEnv{sessionStoreStorage} stateM uriM = do
+    let sessionStoreGenerate = do
+            rb <- liftIO genRandomB64
+            let s = OIDCState (decodeUtf8 rb) uriM
+            pure (B64.encode $ BSL.toStrict $ Data.Aeson.encode s)
+        sessionStoreSave = storeSave
+        sessionStoreGet = storeGet
+        sessionStoreDelete = case stateM of
+            Just state' -> modifyMVar_ sessionStoreStorage $ \store -> pure $ HM.delete state' store
+            Nothing -> pure ()
+     in O.SessionStore{..}
+  where
+    storeSave :: O.State -> O.Nonce -> IO ()
+    storeSave state' nonce = modifyMVar_ sessionStoreStorage $ \store -> pure $ HM.insert state' nonce store
+    storeGet :: O.State -> IO (Maybe O.Nonce)
+    storeGet state' = do
+        store <- readMVar sessionStoreStorage
+        let nonce = HM.lookup state' store
+        pure nonce
+
+loginServer :: OS -> Process -> Sessions -> (Html () -> Html ()) -> JWTSettings -> AuthResult SessionID -> OIDCEnv -> Maybe Workspace -> ServerT LoginAPI Servant.Handler
+loginServer os process sessions mkIndexHtml jwtSettings auth oidcenv mWorkspace =
+    indexRoute auth :<|> loginRoute oidcenv :<|> callbackRoute oidcenv :<|> guestCallbackRoute
+  where
+    callbackRoute :: OIDCEnv -> Maybe Text -> Maybe Text -> Maybe Text -> Servant.Handler AuthResp
+    callbackRoute oidcEnv mErr mCode mState = do
+        case (mErr, mCode, mState) of
+            (Just errorMsg, _, _) -> error $ "Error from remote provider: " <> show errorMsg
+            (_, Nothing, _) -> error "No code parameter given"
+            (_, _, Nothing) -> error "No state parameter given"
+            (_, Just oauthCode, Just oauthState) -> do
+                tokens :: O.Tokens Value <-
+                    liftIO $
+                        O.getValidTokens
+                            (mkSessionStore oidcEnv (Just $ encodeUtf8 oauthState) Nothing)
+                            (oidc oidcEnv)
+                            (manager oidcEnv)
+                            (encodeUtf8 oauthState)
+                            (encodeUtf8 oauthCode)
+                setSession (UserName tokens.idToken.sub) $ SessionProvider "google"
```

Comment:
How about using `SessionProvider ("google": <> tokens.idToken.sub)` here. Then in setSession, we look for such provider. If it is found, then we update the cookie with the attached session. Otherwise we create a new session, using the username from the idp.

So we would need a new `lookupSessionByProvider :: Sessions -> SessionProvider -> ProcessIO (Maybe Session)`.

Reply:
Looking into this some more, it seems like we should enforce unique username, perhaps even for guest. I'll update the Session module to provide a new api.